### PR TITLE
Release scripts for automatic release creation

### PR DESCRIPTION
**Pull request by levitte**
The OpenSSL team has decided to move to automatic release building, to
massively reduce the amount of churn that we go through every time we make a
release.

In total, three scripts will be needed:

1.  a release building script, which we already have two of:

    1. `release-tools/mkrelease.pl` and associated files.
    2. `dev/release.sh` from the source of OpenSSL 3.0 and newer.
  
    This script needs to be able to help build releases manually, at least
    for the moment, as we are moving to automatic release building in steps.
    Security releases are still going to happen manually for the time being,
    and either way, we may want to retain the capability to perform a manual
    release in a pinch.

2.  a script to drive automatic release building.  This script will use the
    release building script, but also provide the infrastructure, such as
    staging directories for release files, a staging repository for the
    temporary tags, etc etc.
    This script will appear in a different repository, probably
    openssl/omc-tools.

3.  a script to publish a chosen release.  We already have such a script,
    described in release-tools/DO-RELEASE.md, which will need some
    modifications, or possibly be replaced entirely.
    This script will appear in a different repository, probably
    openssl/omc-tools.

I will attach the plan for automatic releases in a comment.

### CODE REVIEW

**Comment by levitte** [comment]
The plan as promised

[automated-release-process.md](https://github.com/openssl/tools/files/11086726/automated-release-process.md)

**Comment by levitte** [comment]
The remaining things to do here are:

-   Adapting release-tools/release.sh to be helpful for other scripts, as
    well as alternative locations for repositories and what it outputs.
    
    Essentially, this will add a `--porcelain` option, popularized by git,
    and a few other options to be able to tell it where things should go.

-   Testing (I might add a test suite that runs with a fake OpenSSL repo)

**Comment by levitte** [comment]
I believe I've added all necessary new options to release-tools/release.sh.
Time to write the other scripts, and to test it all.

**Comment by levitte** [comment]
Yesterday, I realised that the plan we had was a bit over the top.  So I rewrote it, see attached files.

[automated-release-process-2.0.pdf](https://github.com/openssl/tools/files/11262785/automated-release-process-2.0.pdf)
[automated-release-process-2.0.md](https://github.com/openssl/tools/files/11262786/automated-release-process-2.0.md)

**Comment by levitte** [comment]
Furthermore, I realised that the original `dev/release.sh` (which is now `release-tools/stage-release.sh` in this PR) is all that we need to stage releases, with just a few more adjustments.

To run this, we've added a few builders on https://ci.buildbot.openssl.org.  Currently, the worker doing the work is running on a test box, with everything being local, to be safe until we've worked out the bugs.

As far as this PR is concerned, I think the job is done.  The stuff needed to public a release will be worked on in another PR.

File: release-tools/release-aux/release-state-fn.sh [hunk 1]
```diff
@@ -0,0 +1,214 @@
+#! /bin/sh
+# Copyright 2020-2021 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (Alich-OpenSSL) [inline release-tools/release-aux/release-state-fn.sh:2]: 2020-2021 ?

**Review by t8m — APPROVED** [review]
Approved with the copyright changes

File: release-tools/release-aux/release-state-fn.sh (hunk 1 above)
Comment (levitte) [reply release-tools/release-aux/release-state-fn.sh:2]: Fixed them all

**Comment by levitte** [comment]
Merged

9317aee3f86c0fa9ff15b8abcd97ee363b7b8a5e Import openssl/openssl's dev/release.sh and accompanying files
032ff96434252242765191d57f4705c02999731e release-tools/release-aux/ documentation updates and small fixes
ccfce28770a23203315575cdc335bcf1a18368a5 Adapt release-tools/release-aux/release-version-fn.sh for pre-3.0 versions
46a367d4c03abffa0fadef508cb41e3f342723a3 Adapt release-tools/release.sh for a life outside of the OpenSSL work tree
396d94fb39bc9da55c2e51e7c9e4ef8fd41b48e5 release-tools/release.sh et al: introduce FULL_VERSION and use it
25d343a424573aff08c6a50e96b1bcb442a140cb release-tools/release.sh et al: introduce RELEASE_FILES and use it
6bbdee09f51d628ced7eff15b52cff1b1475b67e Adapt documentation for the new release script, and clean up
e37897305690cb325b287854c679373748b9a665 Add an option to specify the upload address
d4e73143343f65635eb346bb13611423167db02e Adapt release-tools/release.sh for pre-3.0 versions
a67370165dfd4adb6b8d5527cb32fe2ed768a01d When checking for make update changes, only check tracked files
bf2dcecc58573d2c9f14699aed2548d43f48d47e release.sh: Add --quiet and --porcelain
e539db151d36445072d47619af0aa09f9a8a4bb7 release.sh: Also present the uploaded/to be uploaded files with --porcelain
f3c34e6ea555acecb82e4ee34a4911e4773ddf10 release.sh: Add --clean-worktree, --branch_fmt and --tag-fmt
15b8298e594f5677fb91821f4332c72e07bf0d2f Clarify that the release script only stages the release files
73354762e8bc5e31a4c00cf1f735c1b81229ef0a release-tools/stage-release.sh: Add a metadata file among the release files
e860bbd15081381a1b8ef91ba6df669cd78de459 release-tools/stage-release.sh: Add --unsigned
1fce24a09743e599225a96dcfdb4b212f2a37714 release-tools/stage-release.sh: Use 'type -p' and realpath to find $0
b99e62ffc84d6a8b40d3b03a38d78047287cc2df Include the original HEAD in the porcelain output

**Pull request closed without merging**